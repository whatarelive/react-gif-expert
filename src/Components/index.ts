// Archivo de barril para todas las importaciones de la carpeta componentes.
export * from './AddCategory';
export * from './GifGrid';
export * from './GifItem';