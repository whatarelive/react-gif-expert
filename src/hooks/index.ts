export * from './useCategory';
export * from './useFecthGifs';