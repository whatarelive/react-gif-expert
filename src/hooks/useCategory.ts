import { useState } from "react"

export const useCategory = ( newcategory: string ) => {

    // Estado que contiene el listado de las categorias de los gifs.
    const [ categories, setCategories ] = useState( [newcategory] )

    // Funcion que agrega una nueva categoria al estado inicial.
    const onAddCategory = ( newCategory: string ) => {
        
        // Validacion para saber si la nueva categoria no se encuentra del useState.
        if( !categories.includes( newCategory ) ) {
            setCategories([ newCategory, ...categories])
        }
    }

    return {
        categories,
        onAddCategory
    }
}