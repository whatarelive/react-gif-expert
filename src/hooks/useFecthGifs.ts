// Hooks de React
import { useEffect, useState } from "react";
// API Helpers.
import { getGifs } from "../Helpers/getGifs";
// Types
import { GifType } from "../Types/types"


export const useFecthGifs = ( category: string ) => {

    // Hook useState : permite dal un estado del arreglo de elementos.
    const [ images, setImages ] = useState( Array<GifType> );

    const [ isLoading, setIsLoading ] = useState( true );

    // Funcion para recuperar el arreglo de gifs y guardarlos en el useState. 
    const getImages = async( category: string ) => {
        const newImages = await getGifs( category );
        setImages( newImages );
        setIsLoading( false );
    }

    // Hook UseEffect : permite que solo se llame a la funcion cuando halla un cambio en las categorias.
    useEffect( () => {
        getImages( category );

    }, [category])

    return {
        images,
        isLoading,
    }
}
