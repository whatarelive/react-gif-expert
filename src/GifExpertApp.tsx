// Librerias de React.
import {useState} from "react";
// Componentes hijos.
import { AddCategory, GifGrid } from "./Components";


export const GifExpertApp = () => {

    // Estado que contiene el listado de las categorias de los gifs.
    const [ categories, setCategories ] = useState(['One Punch']);

    // Funcion que agrega una nueva categoria al estado inicial.
    const onAddCategory = ( newCategory: string ) => {
        
        // Validacion para saber si la nueva categoria no se encuentra del useState.
        if ( !categories.includes( newCategory ) ) {
            setCategories([ newCategory, ...categories ]);
        }
    }

    // Funcion que crea el listado de elementos.
    const listItem = () => {
        return categories.map( ( category: string ) => {
           return <GifGrid key={ category } category={ category }/>;
        });
    }

    return (
        <>
            <h3>Gif Expert App</h3>
            <AddCategory onNewCategory={ onAddCategory }/>
            <ol>{ listItem() }</ol>
        </>
    );
};