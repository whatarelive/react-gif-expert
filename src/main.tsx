// Librerias de React.
import React from 'react'
import ReactDOM from 'react-dom/client'
// Estilos CSS. 
import "./Styles/index.css";
// Componentes propios. 
import {GifExpertApp} from "./GifExpertApp.tsx";

const root = document.getElementById('root');

if (root instanceof HTMLDivElement) {

    ReactDOM.createRoot(root).render(
        <React.StrictMode>
                <GifExpertApp/>
        </React.StrictMode>
    );
} 