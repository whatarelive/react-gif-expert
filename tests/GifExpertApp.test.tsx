import React from 'react';
import { render, screen } from '@testing-library/react';
import { useCategory, useFecthGifs } from '../src/hooks';
import { GifExpertApp } from '../src/GifExpertApp';

// Creamos Mocks de los hooks necesezarios para renderizar el componente.
jest.mock('../src/hooks/useCategory');
jest.mock('../src/hooks/useFecthGifs');

describe('Pruebas del componente <GifExpertApp/>', () => { 

    // Data Auxiliar.
    const categories = ['One Punch'];
    const gifs = [
        {
            id: 'ABC',
            title: 'Saitama',
            url: 'https://localhost/saitama.jpg'
        },
        {
            id: '123',
            title: 'Goku',
            url: 'https://localhost/goku.jpg'    
        }
    ];

    // Test #1: Sobre la UI del componente cuando se carga toda la data.
    test('debe mostrar todos los elemntos de forma correcta', () => { 

        // Definimos el valor de retorno del mock del customHook useFetchGifs. 
        (useFecthGifs as jest.Mock).mockReturnValue({
            images: gifs,
            isLoading: false
        });

        // Definimos el valor de retorno del mock del customHook useCategory.
        (useCategory as jest.Mock).mockReturnValue({
            categories: categories,
        });

        // Renderizamos el componente y desetructuramos el contenido del render.
        const { container } = render(<GifExpertApp/>);

        // Esperamos que el contenido del render exitan los siguientes textos.
        expect( screen.getByText('One Punch') );
        expect( screen.getByText('Gif Expert App'));
       
        // Creamos una captura del render.
        expect( container ).toMatchSnapshot();
    });
})