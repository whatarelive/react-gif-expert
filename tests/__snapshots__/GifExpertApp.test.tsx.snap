// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas del componente <GifExpertApp/> debe mostrar todos los elemntos de forma correcta 1`] = `
<div>
  <h3>
    Gif Expert App
  </h3>
  <form
    aria-label="form"
  >
    <input
      placeholder="Buscar Gifs"
      type="text"
      value=""
    />
  </form>
  <ol>
    <h3>
      One Punch
    </h3>
    <div
      class="card-grid"
    >
      <div
        class="card-item"
      >
        <img
          alt="Saitama"
          src="https://localhost/saitama.jpg"
        />
        <p>
          Saitama
        </p>
      </div>
      <div
        class="card-item"
      >
        <img
          alt="Goku"
          src="https://localhost/goku.jpg"
        />
        <p>
          Goku
        </p>
      </div>
    </div>
  </ol>
</div>
`;
